bleak/backends/bumble/__init__.py

```diff
@@ -0,0 +1,66 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2024 Victor Chavez
+"""Bumble backend."""
+from enum import Enum
+from typing import Dict, Final, Optional
+
+from bumble.controller import Controller
+from bumble.link import LocalLink
+from bumble.transport import Transport, open_transport
+
+transports: Dict[str, Transport] = {}
+_link: Final = LocalLink()
+
+
+class TransportScheme(Enum):
+    SERIAL = "serial"
+    UDP = "udp"
+    TCP_CLIENT = "tcp-client"
+    TCP_SERVER = "tcp-server"
+    WS_CLIENT = "ws-client"
+    WS_SERVER = "ws-server"
+    PTY = "pty"
+    FILE = "file"
+    VHCI = "vhci"
+    HCI_SOCKET = "hci-socket"
+    USB = "usb"
+    PYUSB = "pyusb"
+    ANDROID_EMULATOR = "android-emulator"
+    ANDROID_NETSIM = "android-netsim"
+    UNIX = "unix"
+
+
+class BumbleTransport:
+    def __init__(self, scheme: TransportScheme, args: Optional[str] = None):
+        """
+        Args:
+            scheme: TransportScheme: The transport scheme supported by bumble
+            args: Optional[str]: The arguments used to initialize the transport.
+                                 See https://google.github.io/bumble/transports/index.html
+        """
+        self.scheme: Final = scheme
+        self.args: Final = args
+
+    def __str__(self):
+        return f"{self.scheme.value}:{self.args}" if self.args else self.scheme.value
+
+
+def get_default_transport() -> BumbleTransport:
+    return BumbleTransport(TransportScheme.TCP_SERVER, "127.0.0.1:1234")
+
+
+async def start_transport(transport: BumbleTransport) -> None:
+    transport_cmd = str(transport)
+    if transport_cmd not in transports.keys():
+        transports[transport_cmd] = await open_transport(transport_cmd)
```

Comment:
I appreciate all the work done to make this less error-prone; but I'd like to understand why the user cannot simply import their "open transport function" - like `open_serial_transport`, `open_udp_transport`, etc.

It's taking the burden off of Bleak to maintain. Because the function is awaitable and has side effects, it would likely be passed as a partial or lambda (a closure):

```python

async def start_transport(
    open_transport_func: partial[Coroutine[None, None, Transport]]
) -> Transport:
    await open_transport_func()
    ...

my_transport = partial(open_serial_transport, "my serial transport args")

...

await start_transport(my_transport)
```

The specific transport functions are still poorly typed with serialized strings as arguments, but at least now they could change upstream without Bleak having any responsibility to maintain them.

Reply:
I am using the `open_transport` instead of specific `open_transport_xyz` because all the bumble examples and applications use this function. So I assumed this was the official Bumble way 🐝 .

Reply:
Couldnt we map the possible transports functions to an Enum, then the user does not need to pass a partial or callback but just an enum plus string arguments.

Reply:
Yes, that would be cool, but still requires maintenance here when upstream changes.  The closure is perhaps confusing, and an alternative would be that start_transport takes the open_transport_func and spec as args.  If one of the open transport functions upstream changes its number of args, this could be a problem, but it feels like they are using serialized strings with the goal of the function signature being general.

Reply:
My intention with the current implementation is to reuse `open_transport` with the string syntax used in bumble's examples and applications. This provides familiarity for bumble users. The specific string syntax for each transport is documented [here](https://google.github.io/bumble/transports). If I tell the user to import the `open_transport_xyz`, additional documentation needs to be written telling them to look in the source of [`bumble/transport`](https://github.com/google/bumble/tree/main/bumble/transport) and list which functions are available.

I understand your point about upstream changes. Bumble is an experimental project, so there is no guarantee of stability. Since it is still a work in progress, I could also keep this backend as an external module to avoid maintenance costs due to the maturity of the bumble project.

Reply:
> If I tell the user to import the `open_transport_xyz`, additional documentation needs to be written telling them to look in the source of [`bumble/transport`](https://github.com/google/bumble/tree/main/bumble/transport) and list which functions are available.

That's a good point and I agree it's a good justification to use the serialized strings!